File: backend/behavior.py
from __future__ import annotations

from datetime import datetime, timedelta, timezone


def _plan(record: dict) -> dict:
    return record.get("plan") if isinstance(record.get("plan"), dict) else record


def _cost(record: dict) -> float:
    try:
        return max(0.0, float(_plan(record).get("cost", 0)))
    except (TypeError, ValueError):
        return 0.0


def _saved_at(record: dict) -> datetime | None:
    try:
        return datetime.fromisoformat(str(record.get("saved_at", "")).replace("Z", "+00:00"))
    except (TypeError, ValueError):
        return None


def _escalation_count(records: list[dict]) -> int:
    costs = [_cost(record) for record in sorted(records, key=lambda item: str(item.get("saved_at", "")))]
    return sum(1 for index in range(2, len(costs)) if costs[index - 2] < costs[index - 1] < costs[index])


def _market_pulse(snapshot: dict) -> dict:
    rows = [
        {"issue": issue, **item}
        for issue, item in (snapshot.get("issues") or {}).items()
        if isinstance(item.get("sales"), (int, float)) and item["sales"] > 0
    ][-20:]
    if not rows:
        return {
            "available": False,
            "label": "暂无官方销售额",
            "message": "当前数据源没有可用销售额，不以论坛讨论量代替真实市场数据。",
            "source": snapshot.get("source", ""),
            "series": [],
        }
    recent = rows[-5:]
    previous = rows[-10:-5]
    recent_avg = sum(row["sales"] for row in recent) / len(recent)
    previous_avg = sum(row["sales"] for row in previous) / len(previous) if previous else recent_avg
    change = round((recent_avg - previous_avg) / previous_avg * 100, 1) if previous_avg else 0.0
    label = "参与升温" if change >= 5 else "参与降温" if change <= -5 else "参与平稳"
    return {
        "available": True,
        "label": label,
        "change_percent": change,
        "latest_sales": rows[-1]["sales"],
        "latest_issue": rows[-1]["issue"],
        "source": rows[-1].get("source") or snapshot.get("source", ""),
        "message": "官方销售额只描述市场参与规模，不改变任何一注的中奖概率。",
        "series": [{"issue": row["issue"], "sales": row["sales"]} for row in rows],
    }


def build_behavior_profile(records: list[dict], review: dict, snapshot: dict, now: datetime | None = None) -> dict:
    now = now or datetime.now(timezone.utc)
    cutoff = now - timedelta(days=30)
    recent = [record for record in records if (_saved_at(record) or now) >= cutoff]
    costs = [_cost(record) for record in records]
    recent_cost = sum(_cost(record) for record in recent)
    dantuo_count = sum(1 for record in records if _plan(record).get("mode") == "dantuo")
    escalations = _escalation_count(records)
    max_cost = max(costs, default=0)
    avg_cost = round(sum(costs) / len(costs), 2) if costs else 0
    risk_score = min(100, escalations * 30 + (20 if len(recent) >= 8 else 0) + (20 if avg_cost and max_cost >= avg_cost * 2 else 0))
    risk_level = "高" if risk_score >= 60 else "中" if risk_score >= 30 else "低"
    summary = review.get("summary") or {}
    signals = []
    if escalations:
        signals.append(f"发现 {escalations} 次连续三期加码")
    if len(recent) >= 8:
        signals.append("近30日保存方案频次偏高")
    if avg_cost and max_cost >= avg_cost * 2:
        signals.append("最大单次支出达到平均支出的两倍")
    if not signals:
        signals.append("未发现连续加码或异常高频保存")
    action = (
        "暂停新增方案，恢复基础预算并等待已保存方案完成复盘。"
        if risk_level == "高"
        else "保持固定预算，不因市场热度、遗漏或未中奖临时加码。"
    )
    return {
        "engine": "可解释规则引擎",
        "generated_at": now.isoformat(),
        "risk_score": risk_score,
        "risk_level": risk_level,
        "metrics": {
            "record_count": len(records),
            "recent_count": len(recent),
            "recent_cost": round(recent_cost, 2),
            "average_cost": avg_cost,
            "maximum_cost": round(max_cost, 2),
            "dantuo_ratio": round(dantuo_count / len(records) * 100, 1) if records else 0,
            "escalation_count": escalations,
            "reviewed_count": summary.get("reviewed", 0),
            "historical_roi": summary.get("roi") if summary.get("roi_complete") else None,
        },
        "signals": signals,
        "action": action,
        "market": _market_pulse(snapshot),
        "disclaimer": "智能分析只解释历史投注行为与市场参与规模，不预测开奖号码或收益。",
    }

File: frontend/src/behavior.js
function plan(record) {
  return record?.plan || record || {};
}

function cost(record) {
  const value = Number(plan(record).cost || 0);
  return Number.isFinite(value) ? Math.max(0, value) : 0;
}

function escalationCount(records) {
  const costs = [...records]
    .sort((left, right) => Date.parse(left.saved_at || 0) - Date.parse(right.saved_at || 0))
    .map(cost);
  return costs.reduce((count, value, index) => count + (index >= 2 && costs[index - 2] < costs[index - 1] && costs[index - 1] < value ? 1 : 0), 0);
}

function marketPulse(issues = {}, source = "") {
  const rows = Object.entries(issues)
    .map(([issue, item]) => ({ issue, ...item }))
    .filter((item) => Number(item.sales) > 0)
    .slice(-20);
  if (!rows.length) return { available: false, label: "暂无官方销售额", source, message: "当前数据源没有可用销售额，不以论坛讨论量代替真实市场数据。", series: [] };
  const recent = rows.slice(-5);
  const previous = rows.slice(-10, -5);
  const average = (items) => items.reduce((sum, item) => sum + Number(item.sales), 0) / Math.max(1, items.length);
  const recentAverage = average(recent);
  const previousAverage = previous.length ? average(previous) : recentAverage;
  const change = previousAverage ? Math.round(((recentAverage - previousAverage) / previousAverage) * 1000) / 10 : 0;
  return {
    available: true,
    label: change >= 5 ? "参与升温" : change <= -5 ? "参与降温" : "参与平稳",
    change_percent: change,
    latest_sales: Number(rows.at(-1).sales),
    latest_issue: rows.at(-1).issue,
    source: rows.at(-1).source || source,
    message: "官方销售额只描述市场参与规模，不改变任何一注的中奖概率。",
    series: rows.map((item) => ({ issue: item.issue, sales: Number(item.sales) })),
  };
}

export function buildBehaviorProfile(records = [], review = {}, snapshot = {}, now = new Date()) {
  const cutoff = now.getTime() - 30 * 24 * 60 * 60 * 1000;
  const recent = records.filter((record) => {
    const savedAt = Date.parse(record.saved_at || "");
    return !Number.isFinite(savedAt) || savedAt >= cutoff;
  });
  const costs = records.map(cost);
  const averageCost = costs.length ? Math.round((costs.reduce((sum, value) => sum + value, 0) / costs.length) * 100) / 100 : 0;
  const maximumCost = Math.max(0, ...costs);
  const escalations = escalationCount(records);
  const highFrequency = recent.length >= 8;
  const outlier = averageCost > 0 && maximumCost >= averageCost * 2;
  const riskScore = Math.min(100, escalations * 30 + (highFrequency ? 20 : 0) + (outlier ? 20 : 0));
  const riskLevel = riskScore >= 60 ? "高" : riskScore >= 30 ? "中" : "低";
  const signals = [];
  if (escalations) signals.push(`发现 ${escalations} 次连续三期加码`);
  if (highFrequency) signals.push("近30日保存方案频次偏高");
  if (outlier) signals.push("最大单次支出达到平均支出的两倍");
  if (!signals.length) signals.push("未发现连续加码或异常高频保存");
  const summary = review.summary || {};
  return {
    engine: "可解释规则引擎",
    risk_score: riskScore,
    risk_level: riskLevel,
    metrics: {
      record_count: records.length,
      recent_count: recent.length,
      recent_cost: recent.reduce((sum, record) => sum + cost(record), 0),
      average_cost: averageCost,
      maximum_cost: maximumCost,
      dantuo_ratio: records.length ? Math.round((records.filter((record) => plan(record).mode === "dantuo").length / records.length) * 1000) / 10 : 0,
      escalation_count: escalations,
      reviewed_count: summary.reviewed || 0,
      historical_roi: summary.roi_complete ? summary.roi : null,
    },
    signals,
    action: riskLevel === "高" ? "暂停新增方案，恢复基础预算并等待已保存方案完成复盘。" : "保持固定预算，不因市场热度、遗漏或未中奖临时加码。",
    market: marketPulse(snapshot.issues, snapshot.source),
    disclaimer: "智能分析只解释历史投注行为与市场参与规模，不预测开奖号码或收益。",
  };
}
